File: flutter/fastory_sdk/ios/Classes/FastorySdkPlugin.swift
import Flutter
import UIKit

public final class FastorySdkPlugin: NSObject, FlutterPlugin {
    private var eventSink: FlutterEventSink?

    public static func register(with registrar: FlutterPluginRegistrar) {
        let instance = FastorySdkPlugin()
        let methodChannel = FlutterMethodChannel(
            name: "fastory_sdk",
            binaryMessenger: registrar.messenger()
        )
        registrar.addMethodCallDelegate(instance, channel: methodChannel)
        let eventChannel = FlutterEventChannel(
            name: "fastory_sdk/events",
            binaryMessenger: registrar.messenger()
        )
        eventChannel.setStreamHandler(instance)
        Fastory.eventsDelegate = instance
    }

    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch call.method {
        case "configure":
            configure(call, result: result)
        case "openGames":
            openGames(result: result)
        case "close":
            Fastory.close()
            result(nil)
        case "identify":
            identify(call, result: result)
        case "logout":
            Fastory.logout { result(nil) }
        case "setBridgeReply":
            setBridgeReply(call, result: result)
        default:
            result(FlutterMethodNotImplemented)
        }
    }

    private func setBridgeReply(_ call: FlutterMethodCall, result: FlutterResult) {
        let arguments = call.arguments as? [String: Any] ?? [:]
        let wireName = arguments["type"] as? String
        // The registry is re-applied here for the same reason as `configure`: a host can drive the
        // channel directly, and a type outside it must not become a reply the SDK serves.
        guard let type = wireName.flatMap(FastoryBridgeRequestType.init(rawValue:)) else {
            result(FlutterError(
                code: "invalid_bridge_request_type",
                message: "unknown bridge request type: \(wireName ?? "nil")",
                details: nil
            ))
            return
        }
        // Absent or null clears; anything else that is not an object is a host error, never a clear.
        // `as? [String: Any]` alone would map a string or a number to nil and sign the fan out.
        let rawPayload = arguments["payload"]
        var payload: [String: Any]?
        if let rawPayload, !(rawPayload is NSNull) {
            guard let object = rawPayload as? [String: Any] else {
                result(Self.invalidPayload)
                return
            }
            payload = object
        }
        guard FastoryBridgeResponder.setReply(for: type, payload: payload) else {
            result(Self.invalidPayload)
            return
        }
        result(nil)
    }

    private static var invalidPayload: FlutterError {
        FlutterError(
            code: "invalid_bridge_reply_payload",
            message: "the bridge reply payload must be a JSON-serialisable object, or null to clear it",
            details: nil
        )
    }

    private func identify(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        let arguments = call.arguments as? [String: Any] ?? [:]
        let wireName = arguments["mode"] as? String
        guard let mode = wireName.flatMap(FastoryIdentityMode.init(rawValue:)) else {
            result(FlutterError(
                code: "invalid_identity_mode",
                message: "unknown identify mode: \(wireName ?? "nil")",
                details: nil
            ))
            return
        }
        let requested: FastoryIdentity
        switch mode {
        case .anonymous:
            requested = .anonymous
        case .fanId:
            requested = .fanId
        case .hostToken:
            // A blank jwt is rejected by the core as invalid_host_token, not silently accepted.
            requested = .hostToken(jwt: arguments["jwt"] as? String ?? "")
        }
        Fastory.identify(requested) { outcome in
            switch outcome {
            case .success(let identity):
                let resolved: [String: Any] = [
                    "mode": identity.mode.rawValue,
                    "fanId": identity.fanId ?? NSNull(),
                ]
                result(resolved)
            case .failure(let error):
                result(FlutterError(code: error.code, message: "\(error)", details: nil))
            }
        }
    }

    private func configure(_ call: FlutterMethodCall, result: FlutterResult) {
        guard let arguments = call.arguments as? [String: Any] else {
            result(FlutterError(code: "invalid_config", message: "arguments are required", details: nil))
            return
        }
        // Exactly one identifier, enforced by FastoryConfig.validate() below — Dart already
        // rejects the bad combinations, this covers a host calling the channel directly.
        let publishableKey = arguments["publishableKey"] as? String
        let fanzoneSlug = arguments["fanzoneSlug"] as? String
        let environment: FastoryEnvironment
        switch arguments["environment"] as? String {
        case "staging":
            environment = .staging
        case "development":
            guard let developmentBaseUrl = arguments["developmentBaseUrl"] as? String,
                  let baseURL = URL(string: developmentBaseUrl) else {
                result(FlutterError(
                    code: "invalid_config",
                    message: "developmentBaseUrl is required when environment is development",
                    details: nil
                ))
                return
            }
            environment = .development(baseURL: baseURL)
        default:
            environment = .production
        }
        let hubTabSlug = arguments["hubTabSlug"] as? String ?? "games"
        let theme: FastoryTheme? = (arguments["theme"] as? String).flatMap(FastoryTheme.init(rawValue:))
        let config = FastoryConfig(
            environment: environment,
            publishableKey: publishableKey,
            fanzoneSlug: fanzoneSlug,
            hubTabSlug: hubTabSlug,
            locale: arguments["locale"] as? String,
            theme: theme
        )
        do {
            try config.validate()
        } catch {
            result(FlutterError(code: "invalid_config", message: "\(error)", details: nil))
            return
        }
        Fastory.configure(config)
        result(nil)
    }

    private func openGames(result: FlutterResult) {
        guard Fastory.config != nil else {
            result(FlutterError(
                code: "not_configured",
                message: "Fastory.configure() must be called before openGames()",
                details: nil
            ))
            return
        }
        guard let presenter = Self.topViewController() else {
            result(FlutterError(
                code: "no_view_controller",
                message: "openGames() requires a visible view controller",
                details: nil
            ))
            return
        }
        Fastory.openGames(from: presenter)
        result(nil)
    }

    private static func topViewController() -> UIViewController? {
        let keyWindow = UIApplication.shared.connectedScenes
            .compactMap { $0 as? UIWindowScene }
            .flatMap(\.windows)
            .first(where: \.isKeyWindow)
        var top = keyWindow?.rootViewController
        while let presented = top?.presentedViewController {
            top = presented
        }
        return top
    }
}

extension FastorySdkPlugin: FlutterStreamHandler {
    public func onListen(
        withArguments arguments: Any?,
        eventSink events: @escaping FlutterEventSink
    ) -> FlutterError? {
        eventSink = events
        return nil
    }

    public func onCancel(withArguments arguments: Any?) -> FlutterError? {
        eventSink = nil
        return nil
    }
}

extension FastorySdkPlugin: FastoryEventsDelegate {
    public func fastoryHubOpened(fanzoneSlug: String) {
        emit(["type": "hubOpened", "slug": fanzoneSlug])
    }

    public func fastoryHubClosed() {
        emit(["type": "hubClosed"])
    }

    public func fastoryGameOpened(slug: String) {
        emit(["type": "gameOpened", "slug": slug])
    }

    public func fastoryGameClosed() {
        emit(["type": "gameClosed"])
    }

    public func fastoryExternalLink(url: URL) {
        emit(["type": "externalLink", "url": url.absoluteString])
    }

    public func fastoryBridgeMessage(type: String, payload: [String: Any]) {
        // `bridgeType`, not `type`: the channel's own `type` is the event discriminator (SPEC §13.4).
        emit(["type": "bridgeMessage", "bridgeType": type, "payload": payload])
    }

    public func fastoryIdentityResolved(_ identity: FastoryResolvedIdentity) {
        emit([
            "type": "identityResolved",
            "mode": identity.mode.rawValue,
            "fanId": identity.fanId ?? NSNull(),
        ])
    }

    public func fastorySurfaceLoadFailed(_ failure: FastoryLoadFailure) {
        emit([
            "type": "surfaceLoadFailed",
            "surface": failure.surface.rawValue,
            "reason": failure.reason.rawValue,
            "code": failure.code ?? NSNull(),
            "statusCode": failure.statusCode ?? NSNull(),
        ])
    }

    private func emit(_ event: [String: Any]) {
        eventSink?(event)
    }
}

File: flutter/fastory_sdk/android/src/main/kotlin/com/fastory/sdk/flutter/FastorySdkPlugin.kt
package com.fastory.sdk.flutter

import android.app.Activity
import android.content.Context
import com.fastory.sdk.Fastory
import com.fastory.sdk.FastoryBridgeRequestType
import com.fastory.sdk.FastoryBridgeResponder
import com.fastory.sdk.FastoryConfig
import com.fastory.sdk.FastoryEnvironment
import com.fastory.sdk.FastoryEventsListener
import com.fastory.sdk.FastoryIdentifyException
import com.fastory.sdk.FastoryIdentity
import com.fastory.sdk.FastoryIdentityMode
import com.fastory.sdk.FastoryLoadFailure
import com.fastory.sdk.FastoryResolvedIdentity
import com.fastory.sdk.FastoryTheme
import com.fastory.sdk.WorkspaceResolver
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.embedding.engine.plugins.activity.ActivityAware
import io.flutter.embedding.engine.plugins.activity.ActivityPluginBinding
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel

class FastorySdkPlugin :
    FlutterPlugin,
    ActivityAware,
    MethodChannel.MethodCallHandler,
    EventChannel.StreamHandler {

    private var methodChannel: MethodChannel? = null
    private var eventChannel: EventChannel? = null
    private var eventSink: EventChannel.EventSink? = null
    private var activity: Activity? = null
    private var applicationContext: Context? = null

    private val eventsListener = object : FastoryEventsListener {
        override fun onHubOpened(fanzoneSlug: String) =
            emit(mapOf("type" to "hubOpened", "slug" to fanzoneSlug))
        override fun onHubClosed() = emit(mapOf("type" to "hubClosed"))
        override fun onGameOpened(slug: String) = emit(mapOf("type" to "gameOpened", "slug" to slug))
        override fun onGameClosed() = emit(mapOf("type" to "gameClosed"))
        override fun onExternalLink(url: String) = emit(mapOf("type" to "externalLink", "url" to url))
        // `bridgeType`, not `type`: the channel's own `type` is the event discriminator (SPEC §13.4).
        override fun onBridgeMessage(type: String, payload: Map<String, Any?>) =
            emit(mapOf("type" to "bridgeMessage", "bridgeType" to type, "payload" to payload))
        override fun onIdentityResolved(identity: FastoryResolvedIdentity) = emit(
            mapOf(
                "type" to "identityResolved",
                "mode" to identity.mode.wireName,
                "fanId" to identity.fanId,
            ),
        )
        override fun onSurfaceLoadFailed(failure: FastoryLoadFailure) = emit(
            mapOf(
                "type" to "surfaceLoadFailed",
                "surface" to failure.surface.wireName,
                "reason" to failure.reason.wireName,
                "code" to failure.code,
                "statusCode" to failure.statusCode,
            ),
        )
    }

    override fun onAttachedToEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        applicationContext = binding.applicationContext
        methodChannel = MethodChannel(binding.binaryMessenger, "fastory_sdk").also {
            it.setMethodCallHandler(this)
        }
        eventChannel = EventChannel(binding.binaryMessenger, "fastory_sdk/events").also {
            it.setStreamHandler(this)
        }
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        methodChannel?.setMethodCallHandler(null)
        methodChannel = null
        eventChannel?.setStreamHandler(null)
        eventChannel = null
        eventSink = null
        applicationContext = null
    }

    override fun onAttachedToActivity(binding: ActivityPluginBinding) {
        activity = binding.activity
    }

    override fun onDetachedFromActivityForConfigChanges() {
        activity = null
    }

    override fun onReattachedToActivityForConfigChanges(binding: ActivityPluginBinding) {
        activity = binding.activity
    }

    override fun onDetachedFromActivity() {
        activity = null
    }

    override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
        when (call.method) {
            "configure" -> configure(call, result)
            "openGames" -> openGames(result)
            "close" -> {
                Fastory.close()
                result.success(null)
            }
            "identify" -> identify(call, result)
            "logout" -> Fastory.logout { result.success(null) }
            "setBridgeReply" -> setBridgeReply(call, result)
            else -> result.notImplemented()
        }
    }

    override fun onListen(arguments: Any?, events: EventChannel.EventSink?) {
        eventSink = events
    }

    override fun onCancel(arguments: Any?) {
        eventSink = null
    }

    private fun configure(call: MethodCall, result: MethodChannel.Result) {
        val environment = when (call.argument<String>("environment")) {
            "staging" -> FastoryEnvironment.STAGING
            "development" -> FastoryEnvironment.DEVELOPMENT
            else -> FastoryEnvironment.PRODUCTION
        }
        val theme = when (call.argument<String>("theme")) {
            "light" -> FastoryTheme.LIGHT
            "dark" -> FastoryTheme.DARK
            else -> null
        }
        val config = try {
            // Exactly one identifier — FastoryConfig's init enforces it. Dart already rejects the
            // bad combinations; this covers a host calling the channel directly.
            FastoryConfig(
                fanzoneSlug = call.argument<String>("fanzoneSlug"),
                environment = environment,
                hubTabSlug = call.argument<String>("hubTabSlug") ?: "games",
                locale = call.argument<String>("locale"),
                developmentBaseUrl = call.argument<String>("developmentBaseUrl"),
                publishableKey = call.argument<String>("publishableKey"),
                theme = theme,
            )
        } catch (error: IllegalArgumentException) {
            result.error("invalid_config", error.message, null)
            return
        }
        // The key only bootstraps for the applications it was created with, so the resolver needs
        // the host's package name before configure() kicks the exchange off.
        WorkspaceResolver.applicationId = applicationContext?.packageName
        Fastory.configure(config, eventsListener)
        applicationContext?.let(Fastory::preloadHub)
        result.success(null)
    }

    private fun openGames(result: MethodChannel.Result) {
        val currentActivity = activity
        if (currentActivity == null) {
            result.error("no_activity", "openGames() requires a foreground activity", null)
            return
        }
        try {
            Fastory.openGames(currentActivity)
            result.success(null)
        } catch (error: IllegalStateException) {
            result.error("not_configured", error.message, null)
        }
    }

    private fun setBridgeReply(call: MethodCall, result: MethodChannel.Result) {
        val wireName = call.argument<String>("type")
        // The registry is re-applied here for the same reason as `configure`: a host can drive the
        // channel directly, and a type outside it must not become a reply the SDK serves.
        val type = FastoryBridgeRequestType.entries.firstOrNull { it.value == wireName }
        if (type == null) {
            result.error(
                "invalid_bridge_request_type",
                "unknown bridge request type: $wireName",
                null,
            )
            return
        }
        // Absent or null clears; anything else that is not a map is a host error, never a clear. A
        // typed `argument<Map<…>>` would instead throw a ClassCastException here, where iOS silently
        // cleared — the two channels have to refuse the same input the same way.
        val rawPayload = call.argument<Any>("payload")
        if (rawPayload != null && rawPayload !is Map<*, *>) {
            result.error(INVALID_PAYLOAD, INVALID_PAYLOAD_MESSAGE, null)
            return
        }
        if (!FastoryBridgeResponder.setReply(type, rawPayload as Map<*, *>?)) {
            result.error(INVALID_PAYLOAD, INVALID_PAYLOAD_MESSAGE, null)
            return
        }
        result.success(null)
    }

    private fun identify(call: MethodCall, result: MethodChannel.Result) {
        val wireName = call.argument<String>("mode")
        val mode = FastoryIdentityMode.fromWireName(wireName)
        if (mode == null) {
            result.error(
                FastoryIdentifyException.INVALID_IDENTITY_MODE,
                "unknown identify mode: $wireName",
                null,
            )
            return
        }
        val requested = when (mode) {
            FastoryIdentityMode.ANONYMOUS -> FastoryIdentity.Anonymous
            FastoryIdentityMode.FAN_ID -> FastoryIdentity.FanId
            // A blank jwt is rejected by the core as invalid_host_token, not silently accepted.
            FastoryIdentityMode.HOST_TOKEN ->
                FastoryIdentity.HostToken(call.argument<String>("jwt") ?: "")
        }
        try {
            Fastory.identify(requested) { outcome ->
                outcome.fold(
                    onSuccess = {
                        result.success(mapOf("mode" to it.mode.wireName, "fanId" to it.fanId))
                    },
                    onFailure = { error ->
                        val code = (error as? FastoryIdentifyException)?.code
                            ?: FastoryIdentifyException.NETWORK_ERROR
                        result.error(code, error.message, null)
                    },
                )
            }
        } catch (error: IllegalStateException) {
            result.error("not_configured", error.message, null)
        }
    }

    private fun emit(event: Map<String, Any?>) {
        eventSink?.success(event)
    }

    private companion object {
        const val INVALID_PAYLOAD = "invalid_bridge_reply_payload"
        const val INVALID_PAYLOAD_MESSAGE =
            "the bridge reply payload must be a JSON-encodable map, or null to clear it"
    }
}
